### PR TITLE
[NTH-95] Signup

### PR DESCRIPTION
**Pull request by ahmedmurtaja**
- Create Register Service to Register Userts(therapists, clients)
- send email to therapist 
- validate data
- hashing password

### CODE REVIEW

File: src/helpers/validation/userRegister.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import * as Yup from 'yup';
+
+const userRegisterSchema = Yup.object().shape({
+  role: Yup.string().required('Role is required').oneOf(['user', 'therapist'], 'Role must be either user or therapist'),
+  fullName: Yup.string().required('Full name is required'),
+  email: Yup.string().email('Invalid email address').required('Email is required'),
+  password: Yup.string()
+    .required('Password is required')
+    .min(8, 'Password must be at least 8 characters long'),
+  confirmPassword: Yup.string().required('Confirm password is required').oneOf([Yup.ref('password')], 'Passwords must match'),
```
Comment (LinaYahya) [inline src/helpers/validation/userRegister.ts:10]: most of the time no need to send it to the server it's a client validation process to make sure that the user entered the intended password

File: src/controllers/auth.ts [hunk 2]
```diff
@@ -50,4 +53,48 @@ const login = async (req: Request, res: Response, next: NextFunction) => {
...
+      await mailer({
+        to: user.email,
+        subject: 'Account Activation',
+        text: emailText,
+        html: emailBody,
+      });
+      return res.json({
+        message: 'Therapist registered successfully , please check your email for more details',
+      });
+    }
```
Comment (LinaYahya) [inline src/controllers/auth.ts:87]: else  {
...
}
just for readability purposes

File: src/controllers/auth.ts [hunk 3]
```diff
@@ -50,4 +53,48 @@ const login = async (req: Request, res: Response, next: NextFunction) => {
...
+          intro: 'Welcome to Ntherapy! We\'re very excited to have you on board. \n we will review your application and get back to you as soon as possible.',
+          outro: 'Need help, or have questions? Just reply to this email, we\'d love to help.',
+        },
+      };
+      const emailBody = mailGenerator.generate(email);
+      const emailText = mailGenerator.generatePlaintext(email);
+      await mailer({
+        to: user.email,
+        subject: 'Account Activation',
+        text: emailText,
```
Comment (LinaYahya) [inline src/controllers/auth.ts:81]: Not sure if both text and html are required, I don't assume that.
@ahmedmurtaja

File: src/controllers/auth.ts [hunk 4]
```diff
@@ -50,4 +53,48 @@ const login = async (req: Request, res: Response, next: NextFunction) => {
...
 };
 
-export default login;
+const register = async (req: Request, res: Response, next: NextFunction) => {
+  try {
+    const { body } = req;
+    await userRegisterSchema.validate(body);
+    if (body.role === 'therapist') {
+      const user = await registerTherapist(body);
+      const mailGenerator = new Mailgen({
```
Comment (LinaYahya) [inline src/controllers/auth.ts:62]: why not do the email stuff in a separate service

File: src/controllers/auth.ts [hunk 5]
```diff
@@ -50,4 +53,48 @@ const login = async (req: Request, res: Response, next: NextFunction) => {
...
+      };
+      const emailBody = mailGenerator.generate(email);
+      const emailText = mailGenerator.generatePlaintext(email);
+      await mailer({
+        to: user.email,
+        subject: 'Account Activation',
+        text: emailText,
+        html: emailBody,
+      });
+      return res.json({
```
Comment (LinaYahya) [inline src/controllers/auth.ts:84]:
```suggestion
      return res.status(201).json({
```

File: src/controllers/auth.ts [hunk 6]
```diff
@@ -50,4 +53,48 @@ const login = async (req: Request, res: Response, next: NextFunction) => {
...
+  try {
+    const { body } = req;
+    await userRegisterSchema.validate(body);
+    if (body.role === 'therapist') {
+      const user = await registerTherapist(body);
+      const mailGenerator = new Mailgen({
+        theme: 'salted',
+        product: {
+          name: 'Ntherapy',
+          link: 'https://ntherapy.com/',
```
Comment (LinaYahya) [inline src/controllers/auth.ts:66]: can we add this to config file and import it as an environment variable

**Review by LinaYahya — CHANGES_REQUESTED** [review]

File: src/controllers/auth.ts [hunk 7]
```diff
@@ -50,4 +53,48 @@ const login = async (req: Request, res: Response, next: NextFunction) => {
   }
 };
 
-export default login;
+const register = async (req: Request, res: Response, next: NextFunction) => {
+  try {
+    const { body } = req;
+    await userRegisterSchema.validate(body);
+    if (body.role === 'therapist') {
```
Comment (LinaYahya) [inline src/controllers/auth.ts:60]:
```suggestion
    if (body.role === RolesForSelected.therapist) {
```

**Review by LinaYahya — CHANGES_REQUESTED** [review]

File: src/controllers/auth.ts (hunk 2 above)
Comment (ahmedmurtaja) [reply src/controllers/auth.ts:87]: it gives me eslint error no-else-return

Unnecessary 'else' after 'return'.

File: src/services/auth.ts [hunk 8]
```diff
@@ -10,5 +13,30 @@ const loginByEmail = async (email: string) => {
...
+const registerUser = async ({
+  fullName, email, password, role, phoneNumber,
+}: IUser) => {
+  const userExist = await User.findOne({ where: { email } });
+  if (userExist) {
+    throw templateErrors.BAD_REQUEST('user already exist, please login instead');
+  }
+  const hashedPassword = await bcrypt.hash(password, 12);
+  const user = await User.create({
+    fullName, email, password: hashedPassword, role, phoneNumber,
```
[the remark this answers is not in the document]
Comment (ahmedmurtaja) [reply src/services/auth.ts:25]: done

**Review by LinaYahya — APPROVED** [review]

**Review by MohannadSabaa — APPROVED** [review]

**Review by raghdarizq — APPROVED** [review]

**Review by YasserAmro0 — APPROVED** [review]

**Pull request merged**